deployment/infrastructure/ec2/instance.ts

```diff
@@ -1,21 +1,44 @@
 import * as pulumi from "@pulumi/pulumi";
 import * as aws from "@pulumi/aws";
+import * as fs from "fs";
 
 import { Network } from "./network";
+import { Output } from "@pulumi/pulumi";
 
 export function createInstance(
   instanceType: pulumi.Input<string>,
+  volumeSize: number,
   ami: pulumi.Input<string>,
   network: Network,
-  keyPair: aws.ec2.KeyPair,
+  key: aws.ec2.KeyPair,
+  userDataPath: string = "",
   instanceName = pulumi.getProject()
 ) {
+  let userData;
+  if (userDataPath) {
+    const config = new pulumi.Config();
+    userData = fs
+      .readFileSync(userDataPath, "utf8")
+      .replace(/{{ ([^ ]+) }}/g, (original, matched) => {
```

Comment:
I assume there is a reason why this is here. Could you explain why? Either here or in a comment in the code.

Reply:
this code is used to replace variables `{{ var }}` in user data, in this project it is used to inject ghcr username and password in `deployment/configuration/backend.nix` as I couldn't find any more elegant solution.